File: src/HashTable.rs
use std::fmt::Display;
pub const CAP: usize = 10;

#[derive(Debug, Clone, Copy)]
struct Cell<Key, Value> {
    key: Key,
    value: Value,
    taken: bool
}

impl<Key: Default + Clone, Value: Default + Clone>Cell<Key, Value> {
    fn new() -> Self {
        Self {
            key: Key::default(),
            value: Value::default(),
            taken: false,
        }
    }
}


#[derive(Debug, Clone)]
struct HashTable<Key, Value> { 
    pairs: Vec<Cell<Key, Value>>,
    cap: usize,
    size: usize,
}

trait Hashable {
    fn hash(&self) -> usize;
}



impl Hashable for String {
    fn hash(&self) -> usize {
        let mut hash: usize = 5381;

        for c in self.bytes() {
            hash = ((hash << 5).wrapping_add(hash)).wrapping_add(c as usize);
        }

        hash
    }
}

impl<Key: Default + Clone + Hashable + Display, Value: Default + Clone + Display> HashTable<Key, Value> {
    fn new() -> Self {
        Self {
            pairs: vec![Cell::new(); CAP],
            cap: CAP,
            size: 0,
        }
    }

    fn keys(&mut self) -> Vec<Key> {
        return self.pairs.iter()
            .filter(|cell| cell.taken)
            .map(|cell| cell.key.clone())
            .collect::<Vec<Key>>();
    }
    
    fn values(&mut self) -> Vec<Value> {
        return self.pairs.iter()
            .filter(|cell| cell.taken)
            .map(|cell| cell.value.clone())
            .collect::<Vec<Value>>();
    }
    
    fn expand(&mut self) -> bool {
        
        if self.size == self.cap {
            let mut new_ = vec![Cell::new(); self.cap + CAP];
            
            for i in 0..self.cap {
                new_[i] = self.pairs[i].clone();
            }

            self.cap = self.cap + CAP;
            return true;
        }
        return false;
    }

    fn get(&mut self, key: &Key) -> Option<&Value> {
        let index = key.hash() % CAP;
        
        if self.pairs[index].taken {
            return Some(&self.pairs[index].value);
        }

        return None;
    }

    
    fn get_mut(&mut self, key: &Key) -> Option<&mut Value> {
        let index = key.hash() % CAP;
        
        if self.pairs[index].taken {
            return Some(&mut self.pairs[index].value);
        }

        return None;
    }
    
    fn increment_size(&mut self) {
        self.size = (self.size + 1);
    }

    fn insert(&mut self, key: Key, value: Value) -> bool {
        let mut index = key.hash() % CAP;
    
        while self.pairs[index].taken {
            index = index + 1;
        }

        if index == CAP {
            // TODO: Make the table expandable for more elements.
            return false;
        }
        
        self.pairs[index].key   = key;
        self.pairs[index].value = value;
        self.pairs[index].taken = true;
        self.increment_size();
        return true;
    }
    
    fn delete(&mut self) {
        todo!();
    }
    
    fn contains(&mut self, key: Key) {
        todo!();
    }
    
    fn display(&mut self) {
        for i in 0..self.cap {
            let cell = &self.pairs[i];
            if cell.taken {
                println!("{} -> {}", cell.key, cell.value);
            }
        }
    }
    

    fn display_keys(&mut self) {
        let ks = self.keys();
        for i in 0..self.size {
            println!("{}", ks[i])
        }
    }
    
    fn display_values(&mut self) {
        let values = self.values();
        for i in 0..self.size {
            println!("{}", values[i])
        }
    }
    
    fn display_cells(&mut self) {
        for i in 0..self.cap {
            let cell = &self.pairs[i];
            println!("({}, {}, {})", cell.key, cell.value, cell.taken);
        }
    }
}

fn print_table(table: &HashTable<String, String>) {

    for i in 0..CAP {
        let cell = &table.pairs[i];
        
        if cell.taken {
            println!("{} -> {}", cell.key, cell.value);
            continue;
        }
        
        println!("x");
    }

}

#[allow(unused_variables)]
fn main() { 
    let mut table: HashTable<String, String> = HashTable::new();
    

    table.insert("England".to_string(), "London".to_string()); 
    table.insert("USA".to_string(), "NYC".to_string()); 
    table.insert("JAPAN".to_string(), "Tokyo".to_string());
    println!("----CELLS-----"); 
    table.display_cells();
    println!("----KEYS-----"); 
    table.display_keys();
    println!("----VALUES-----"); 
    table.display_values();

}

File: src/hashtable.c
#include <hashtable.h>

unsigned long djb2_hash(unsigned char *key)
{
	unsigned long hash_ = 5381;
	int c;
	while ((c = *key++))
		hash_ = ((hash_ << 5) + hash_) + c;
	return hash_;
}

void expand_table(HashTable *self) 
{
	self->cells = realloc(self->cells, self->cap * 2 * self->cells[0]);
	self->cap *= 2;
}

bool get(HashTable *self, char *key, Cell *bucket)
{
	
	size_t 	index = djb2_hash((unsigned char *)key) % self->cap;
	while(strcmp(key, self->cells[index].key) != 0) {
		index++;
		if(index == self->cap)
			return false;	
	}
	if (strcmp(key, self->cells[index].key ) == 0) {
		bucket->key   = key;
		bucket->value = self->cells[index].value;
		bucket->taken = self->cells[index].taken;
		return true;
	}
	return false;
}

bool contains(HashTable *self, char *key)
{
	
	size_t 	index  = djb2_hash((unsigned char *)key) % self->cap;
	while(strcmp(key, self->cells[index].key) != 0) {
		index++;

		if(index == self->cap)
			return false;	
		
	}
	if (strcmp(key, self->cells[index].key ) == 0) 
		return true;
	return false;
}

bool insert(HashTable *self, char *key, int value)
{
	bool    exists = false;
	Cell    *cell;
	size_t 	index  = djb2_hash((unsigned char *)key) % self->cap;

	while(self->cells[index].taken)
	{
		if(strcmp(self->cells[index].key, key) == 0)
		{
			exists = true;
			break;
		}

		index++;
		if (index >= self->cap)
			expand_table(self);
	}

	if(index >= self->cap)
		expand_table(self);
	cell = (self->cells + index);
	if (!exists) {
		
		cell->key   = strdup(key);
		cell->value = value;
		cell->taken = true;
		self->size++;
	} else
		cell->value = value;
	return true;
}

HashTable *new_hash_table(void)
{
	HashTable *table = calloc(1, sizeof(*table));
	if (!table)
		return (NULL);
	table->cells = calloc(CAP, sizeof(table->cells[0]));
	if (!table->cells)
	{
		free(table);
		return (NULL);
	}
	table->size = 0;
	table->cap  = CAP;
	return (table);
}

char **TD_alloc(size_t size)
{
	char **TD = (char**) calloc(size, sizeof(char*));
	for(size_t i = 0; i < size; i++)
		TD[i] = (char *) calloc(128, 1);	
	return TD;
}

char **keys(HashTable *self) 
{
	char **ks = TD_alloc(self->size + 1);
	size_t it = 0;

	for(size_t n = 0; n < self->cap; n++)
	{
		if(self->cells[n].taken)
			ks[it++] = self->cells[n].key;
	}
	ks[it++] = NULL;
	return (ks);
}

int *values(HashTable *self) 
{
	int *vs = (int *) calloc(self->size, sizeof(int));
	for(size_t n = 0, it = 0; n < self->cap; n++)
	{
		if(self->cells[n].taken)
			vs[it++] = self->cells[n].value;	
	}
	return vs;
}

void display_cells(HashTable *table) {
	for(size_t i = 0; i < table->cap; i++)
	{
		Cell cell = table->cells[i];
		printf("%zu  -> (%s, %i, %i)\n", 
			i, cell.key, cell.value, cell.taken);
	}
}

void display_keys(HashTable *self) 
{
	char **ks = keys(self);
	for(size_t i = 0; i < self->size; i++)
	{
		printf("%s\n", ks[i]);
	}
}

void display_values(HashTable *self) 
{
	int *vs = values(self);
	
	for(size_t i = 0; i < self->size; i++)
	{
		printf("%i\n", vs[i]);
	}
}
